src/command_controller.rs

```diff
@@ -0,0 +1,288 @@
+//!    );
+//!    Ok(())
+//! }
+//! ```
+//!
+use crate::client::Client;
+use crate::controller::{Controller, ControllerStrategy, ReconciliationState};
+use crate::error::{Error, OperatorResult};
+use crate::reconcile::{ReconcileFunctionAction, ReconcileResult, ReconciliationContext};
+use crate::{metadata, Crd};
+use async_trait::async_trait;
+use k8s_openapi::apimachinery::pkg::apis::meta::v1::OwnerReference;
+use kube::api::{ListParams, Meta};
+use kube::Api;
+use serde::de::DeserializeOwned;
+use std::fmt::Debug;
+use std::future::Future;
+use std::pin::Pin;
+use std::time::Duration;
+
+const FINALIZER_NAME: &str = "command.stackable.tech/cleanup";
+
+/// The Owner trait represents our main controller object. This is required to be able to search
+/// for the main controller object and extract metadata information to set the owner reference
+/// in our command object. This must be implemented by every command custom resource.
+pub trait CommandOwner {
+    type Owner: Meta + Clone + DeserializeOwned + Debug + Send + Sync + Crd;
+    /// Retrieve the potential "Owner" name of this custom resource
+    fn get_owner_name(&self) -> String;
+}
+
+struct CommandState<T> {
+    context: ReconciliationContext<T>,
+}
+
+impl<T> CommandState<T>
+where
+    T: CommandOwner + Clone + DeserializeOwned + Meta,
+{
+    /// This controller sets the owner reference in our custom resource object. We need to
+    /// find the potential owner, extract its metadata into an OwnerReference object and
+    /// patch our command custom resource with that OwnerReference. This makes sure we can
+    /// list and work with the commands in the main controller loop.
+    async fn set_owner_reference(&mut self) -> ReconcileResult<Error> {
+        let owner = find_owner::<T::Owner>(
+            &self.context.client.clone(),
+            &self.context.resource.get_owner_name(),
+        )
+        .await?;
+
+        let owner_reference =
+            metadata::object_to_owner_reference::<T::Owner>(owner.meta().clone(), true)?;
+
+        patch_owner_reference(
+            &self.context.client,
+            &self.context.resource,
+            &owner_reference,
+        )
+        .await?;
+
+        Ok(ReconcileFunctionAction::Done)
+    }
+}
+
+impl<T> ReconciliationState for CommandState<T>
+where
+    T: CommandOwner + Clone + DeserializeOwned + Meta + Send + Sync,
+{
+    type Error = Error;
+
+    fn reconcile(
+        &mut self,
+    ) -> Pin<Box<dyn Future<Output = Result<ReconcileFunctionAction, Self::Error>> + Send + '_>>
+    {
+        Box::pin(async move { self.set_owner_reference().await })
+    }
+}
+
+#[derive(Debug)]
+struct CommandStrategy<T> {
+    // TODO: Better workaround for PhantomData?
+    // We use it here cause we need to make CommandStrategy generic to be able to do:
+    // impl<T> ControllerStrategy for CommandStrategy<T>
+    _ignore: Option<std::marker::PhantomData<T>>,
```

Comment:
This makes me wonder if my architecture with the strategy etc. is not a good one....
I see why you need it and I don't see a good way to remove it.